### PR TITLE
New study mode for Minigui

### PR DESCRIPTION
**Pull request by tommadams**
This basically ended up being another rewrite because performance wasn't good enough.

I'm just throwing this PR out there so that interested parties can clone it. I'll break it up into smaller PRs next week.

### CODE REVIEW

File: cc/mcts_player_test.cc [hunk 1]
```diff
@@ -177,9 +177,15 @@ TEST(MctsPlayerTest, PickMoveArgMax) {
   auto player = CreateBasicPlayer(options);
   auto* root = player->root();
 
-  root->edges[Coord(2, 0)].N = 10;
-  root->edges[Coord(1, 0)].N = 5;
-  root->edges[Coord(3, 0)].N = 1;
+  std::vector<std::pair<Coord, int>> child_visits = {
```
Comment (amj) [inline cc/mcts_player_test.cc]: does Undo or Reset need tests?

File: minigui/serve.py [hunk 2]
```diff
@@ -109,27 +110,46 @@ def _open_pipes():
...
     global token
     global echo_streams
 
-    for line in p.__getattribute__(stream):
+    if echo_streams:
+        sys.stdout.write(line)
+        if "GTP engine ready" in line:
+            echo_streams = False
+
+    if line[-1] == "\n":
```
[the remark this answers is not in the document]
Comment (tommadams) [reply minigui/serve.py]: I want to keep trailing whitespace intact. TBH now that you've got me thinking about it, I probably should do the newline trimming in the frontend. Added a TODO.

File: cc/sgf.cc [hunk 3]
```diff
@@ -191,6 +191,39 @@ class Parser {
   std::string* error_;
 };
 
+void GetMoveTreeImpl(const Ast::Tree& tree,
```
[the remark this answers is not in the document]
Comment (tommadams) [reply cc/sgf.cc]: GetMainLineMoves returns the left-most chain of children. I've added a TODO to remove it.

File: cc/gtp_player.cc [hunk 4]
```diff
@@ -83,32 +127,27 @@ void GtpPlayer::RegisterCmd(const std::string& cmd, CmdHandler handler) {
...
-  TreeSearch();
-
-  ponder_count_ += options().batch_size;
-  if (ponder_count_ >= ponder_limit_) {
-    std::cerr << root()->Describe() << "\n";
-    std::cerr << "finished pondering" << std::endl;
+  if (ponder_read_count_ == 0) {
+    std::cerr << "pondering..." << std::endl;
   }
+  int n = root()->N();
```
[the remark this answers is not in the document]
Comment (tommadams) [reply cc/gtp_player.cc]: Added a comment.

File: cc/gtp_player.cc [hunk 5]
```diff
@@ -490,48 +550,189 @@ GtpPlayer::Response GtpPlayer::HandleReportSearchInterval(absl::string_view cmd,
...
+  }
+
+  if (!UndoMove()) {
+    return Response::Error("cannot undo");
+  }
+
+  return Response::Ok();
+}
+
+GtpPlayer::Response GtpPlayer::HandleVariation(absl::string_view cmd,
```
[the remark this answers is not in the document]
Comment (tommadams) [reply cc/gtp_player.cc]: I guess the .h file.

I'd also like to write a guide to adding Minigui support to an existing go engine once.

File: cc/mcts_player_test.cc (hunk 1 above)
Comment (tommadams) [reply cc/mcts_player_test.cc]: Done.

File: minigui/variation_tree.ts [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+import {getElement, pixelRatio} from './util'
+import {View} from './view'
+
+// Spacing between nodes in the tree.
+const SPACE = 19;
+
+// Padding from the edge of the canvas.
+const PAD = 11;
+
+// Raius of the nodes.
```
[the remark this answers is not in the document]
Comment (tommadams) [reply minigui/variation_tree.ts]: done.

File: cc/mcts_node.cc [hunk 7]
```diff
@@ -106,19 +106,20 @@ MctsNode::MctsNode(MctsNode* parent, Coord move)
...
-  for (int i = 1; i < kNumMoves; ++i) {
+  int best_N = -1;
+  for (int i = 0; i < kNumMoves; ++i) {
     int cn = child_N(i);
-    if (cn > best_N) {
-      moves.clear();
-      moves.push_back(i);
-      best_N = cn;
-    } else if (cn == best_N) {
+    if (cn >= best_N && children.contains(i)) {
```
Comment (sethtroisi) [inline cc/mcts_node.cc]: can the contains check be changed to && cn > 0?

Comment (tommadams) [reply cc/mcts_node.cc]: This is a weird edge case that I'm probably not dealing with in quite the best way. I've added an explanatory comment and a TODO.

File: cc/mcts_player_test.cc [hunk 8]
```diff
@@ -472,6 +480,45 @@ TEST(MctsPlayerTest, SymmetriesTest) {
...
 
+TEST(MctsPlayerTest, ResetRoot) {
+  auto player = absl::make_unique<TestablePlayer>(MctsPlayer::Options());
+
+  auto* game_root = player->root();
+
+  auto c = Coord(12);
+  player->PlayMove(c);
+
+  EXPECT_EQ(c, player->root()->move);
```
[the remark this answers is not in the document]
Comment (tommadams) [reply cc/mcts_player_test.cc]: N will actually be 0 here because the child was never visited during tree search.

File: cc/sgf.cc [hunk 9]
```diff
@@ -191,6 +191,39 @@ class Parser {
...
+                     std::vector<std::unique_ptr<MoveTree>>* dst) {
```
[the remark this answers is not in the document]
Comment (tommadams) [reply cc/sgf.cc]: I don't think that would help. The SGF syntax is a bit weird for a format that describes a tree since most often the game trees have a very low branching factor (~1 in most cases).

A list of nodes in the SGF source corresponds to a chain of descendants not siblings, so the shape of the source SGF tree doesn't actually correspond to the shape of the parsed game tree.

To give an example, a source tree that looks like this:
```
   (a b c)
   /     \
(d e)   (f g)
```

Parses to a game tree that looks like this:
```
(a)
 |
(b)
 |
(c)
 |  \
(d) (f)
 |   |
(e) (g)
```

File: cc/sgf.h [hunk 10]
```diff
@@ -21,6 +21,7 @@
 #include <utility>
 #include <vector>
 
+#include "absl/container/flat_hash_map.h"
```
[the remark this answers is not in the document]
Comment (tommadams) [reply cc/sgf.h]: done.

File: minigui/app.ts [hunk 11]
```diff
@@ -12,103 +12,180 @@
...
     let uri = `http://${document.domain}:${location.port}/minigui`;
-    return this.gtp.connect(uri).then((size: number) => {
-      this.size = size;
+    let params = new URLSearchParams(window.location.search);
+    let p = params.get("gtp_debug");
+    let debug = (p != null) && (p == "" || p == "1" || p.toLowerCase() == "true");
+    return this.gtp.connect(uri, debug).then((size: number) => {
+      setBoardSize(size);
+
+      let stones = new Array<Color>(N * N);
```
[the remark this answers is not in the document]
Comment (tommadams) [reply minigui/app.ts]: Indeed it is. I've added a comment explaining why this is so.

File: minigui/base.ts [hunk 12]
```diff
@@ -27,11 +27,36 @@ function otherColor(color: Color) {
...
+
 enum BoardSize {
   Nine = 9,
   Nineteen = 19,
 }
 
+let N = BoardSize.Nineteen;
+
+function setBoardSize(size: number) {
+  switch (size) {
```
[the remark this answers is not in the document]
Comment (tommadams) [reply minigui/base.ts]: hah, some kind of leftover. thanks for catching that

File: minigui/base.ts [hunk 13]
```diff
@@ -46,6 +71,28 @@ interface Point {
...
+  if (move == 'pass' || move == 'resign') {
+    return move;
+  }
+  let row = N - move.row;
+  let col = COL_LABELS[move.col];
+  return `${col}${row}`;
+}
+
+function movesEqual(a: Nullable<Move>, b: Nullable<Move>) {
+  if (a == null || b == null ||
```
[the remark this answers is not in the document]
Comment (tommadams) [reply minigui/base.ts]: done. thanks, I wrote `moveIsPoint` long after `movesEqual`.

File: minigui/board.ts [hunk 14]
```diff
@@ -188,11 +186,11 @@ class Board extends View {
 
     ctx.fillStyle = this.stoneFill(color, alpha);
 
-    let r = 0.48 * this.stoneRadius;
+    let r = this.stoneRadius;
     for (let p of ps) {
       let c = this.boardToCanvas(p.row, p.col);
       ctx.beginPath();
-      ctx.translate(c.x, c.y);
+      ctx.translate(c.x + 0.5, c.y + 0.5);
```
[the remark this answers is not in the document]
Comment (tommadams) [reply minigui/board.ts]: The 0.5 offsets the center of the circle by half a pixel so that it's correctly centered.

I actually need to do another pass over the all the drawing code to make it consistent about how it handles issues like this.

File: cc/sgf.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (sethtroisi) [reply cc/sgf.cc]: I misunderstood what was happening.
